File: sehun/16week/BOJ_9012.py
def check_ps(input: str) -> None:
    stack = []

    for char in input:
        # 입력이 여는 괄호일 때는 push만 실행됨
        if char == "(":
            stack.append(char)
        # 입력이 닫는 괄호일 때는 pop만 실행되어야 함
        else:
            # stack이 비어있는 경우에 pop이 실행되면 예외가 발생됨
            # 해당케이스 : 「())」의 마지막 「)」
            if stack:
                stack.pop()
            # 스택이 비어있는 경우에 「)」 입력이 오는경우는 ps가 될 수 없음
            else:
                # NO의 경우에 그냥 return해버려서 함수를 종료시키는게 좋은 해결방안인지는 모르겠음
                return print("NO")

    # 최종적으로 stack이 비어있으면 모든 괄호가 짝이 맞은 경우임
    if stack:
        print("NO")
    else:
        print("YES")


t = int(input())

for i in range(t):
    check_ps(input())

File: kimchanho/16week/BOJ_9012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * @title 괄호 (9012)
 * @description 괄호 문자열이 주어졌을 때, 문자열이 올바른 괄호 문자열인지 판단하기
 * @author Chanho Kim
 * @see https://www.acmicpc.net/problem/9012
 */
public class BOJ_9012 {

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(br.readLine());

    for (int i = 0; i < n; i++) {
      char[] input = br.readLine().toCharArray();
      isVPS(input);
    }
  }

  public static void isVPS(char[] input) {
    Deque<Character> stack = new ArrayDeque<>();
    for (char c : input) {
      // 여는 괄호면 스택에 추가
      if (c == '(') {
        stack.push(c);
      // 닫는 괄호면 스택에서 제거
      } else {
        // 스택이 비어있으면 올바르지 않은 괄호 문자열이므로 NO 출력 후 종료
        if (stack.isEmpty()) {
          System.out.println("NO");
          return;
        }
        stack.pop();
      }
    }
    System.out.println(stack.isEmpty() ? "YES" : "NO");
  }
}